In [802]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

# นำเข้าข้อมูล

In [803]:
BNN = pd.read_excel('Banana.xlsx', sheet_name = 'Property')

In [804]:
BNN.index = BNN['ID']
BNN = BNN.drop(['ID', 'Storage type'],  axis = 1)

# สร้าง Model MLR

## จาก ตัวแปร L* a*  และ b* ทำนาย Brix

In [805]:
X1 = BNN[['a*','b*']]
X2 = BNN[['a*','L*']]
X3 = BNN[['L*','b*']]

In [806]:
A1 = BNN[['a*','b*']]
A2 = BNN[['a*','L*']]
A3 = BNN[['L*','b*']]

In [807]:
Y = BNN[['acid']]

In [808]:
B = BNN[['Firmness']]

In [809]:
from sklearn.model_selection import train_test_split

In [810]:
from sklearn.linear_model import LinearRegression

In [811]:
from sklearn import metrics

In [812]:
X1_train, X1_test, y_train, y_test = train_test_split(X1, Y, test_size=0.20, random_state=100)
MLR1 = LinearRegression()
MLR1.fit(X1_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [813]:
X2_train, X2_test, y_train, y_test = train_test_split(X2, Y, test_size=0.20, random_state=100)
MLR2 = LinearRegression()
MLR2.fit(X2_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [814]:
X3_train, X3_test, y_train, y_test = train_test_split(X3, Y, test_size=0.20, random_state=100)
MLR3 = LinearRegression()
MLR3.fit(X3_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [815]:
A1_train, A1_test, B_train, B_test = train_test_split(A1, B, test_size=0.20, random_state=100)
MLR4 = LinearRegression()
MLR4.fit(A1_train, B_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [816]:
A2_train, A2_test, B_train, B_test = train_test_split(A2, B, test_size=0.20, random_state=100)
MLR5 = LinearRegression()
MLR5.fit(A2_train, B_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [817]:
A3_train, A3_test, B_train, B_test = train_test_split(A3, B, test_size=0.20, random_state=100)
MLR6 = LinearRegression()
MLR6.fit(A3_train, B_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [818]:
MLR1.coef_
MLR2.coef_
MLR3.coef_
MLR4.coef_
MLR5.coef_
MLR6.coef_

array([[ 0.01574696, -0.44984019]])

In [819]:
MLR1.intercept_
MLR2.intercept_
MLR3.intercept_
MLR4.intercept_
MLR5.intercept_
MLR6.intercept_

array([20.80230341])

In [820]:
X1_train_prd = MLR1.predict(X1_train)
X2_train_prd = MLR2.predict(X2_train)
X3_train_prd = MLR3.predict(X3_train)

In [821]:
A1_train_prd = MLR4.predict(A1_train)
A2_train_prd = MLR5.predict(A2_train)
A3_train_prd = MLR6.predict(A3_train)

In [822]:
X1_test_prd = MLR1.predict(X1_test)
X2_test_prd = MLR2.predict(X3_test)
X3_test_prd = MLR3.predict(X3_test)

In [823]:
A1_test_prd = MLR4.predict(A1_test)
A2_test_prd = MLR5.predict(A2_test)
A3_test_prd = MLR6.predict(A3_test)

## Brix = -0.261L*+0.857a*+0.294b*+15.160

In [824]:
print('acid')
if MLR1.score(X1_train, y_train)>MLR2.score(X2_train, y_train) and MLR1.score(X1_train, y_train)>MLR3.score(X3_train, y_train):
  print('R\u00b2 of train =', MLR1.score(X1_train, y_train), 'R\u00b2 of test =', MLR1.score(X1_test, y_test))
elif MLR2.score(X2_train, y_train)>MLR1.score(X1_train, y_train) and MLR2.score(X2_train, y_train)>MLR3.score(X3_train, y_train):
  print('R\u00b2 of train =', MLR2.score(X2_train, y_train), 'R\u00b2 of test =', MLR2.score(X2_test, y_test))
elif MLR3.score(X3_train, y_train)>MLR1.score(X1_train, y_train) and MLR3.score(X3_train, y_train)>MLR2.score(X2_train, y_train):
  print('R\u00b2 of train =', MLR3.score(X3_train, y_train), 'R\u00b2 of test =', MLR3.score(X3_test, y_test))
#print('RMSEC =', np.sqrt(metrics.mean_squared_error(y_train, X_train_prd)))
#print('RMSEP =', np.sqrt(metrics.mean_squared_error(y_test, X_test_prd)))
print('\nfirmness')
if MLR4.score(A1_train, B_train)>MLR5.score(A2_train, B_train) and MLR4.score(A1_train, B_train)>MLR6.score(A3_train, B_train):
 print('R\u00b2 of train =', MLR4.score(A1_train, B_train), 'R\u00b2 of test =', MLR4.score(A1_test, B_test))
elif MLR5.score(A2_train, B_train)>MLR4.score(A1_train, B_train) and MLR5.score(A2_train, B_train)>MLR6.score(A3_train, B_train):
  print('R\u00b2 of train =', MLR5.score(A2_train, B_train), 'R\u00b2 of test =', MLR5.score(A2_test, B_test))
elif MLR6.score(A3_train, B_train)>MLR4.score(A1_train, B_train) and MLR6.score(A3_train, B_train)>MLR5.score(A2_train, B_train):
  print('R\u00b2 of train =', MLR6.score(A3_train, B_train), 'R\u00b2 of test =', MLR6.score(A3_test, B_test))

acid
R² of train = 0.7895598938759782 R² of test = 0.8288388983506104

firmness
R² of train = 0.6873857644414216 R² of test = 0.7522069951215451


## การบ้าน
1. ให้นักศึกษา สร้าง Model MLR ทำนาย acid และ firmness โดยทดลองเป็นค่า L* a* b* ไปเรื่อยเพื่อดู ค่า R2 ที่สูงที่สุด

Ans. 
     
     acid = a* , b*
     R² of train = 0.7895598938759782 R² of test = 0.8288388983506104
     RMSEC = 0.030976137047004818
     RMSEP = 0.031121614521102858

     firmness = a* , b*
     R² of train = 0.6873857644414216 R² of test = 0.7522069951215451
     RMSEC = 1.3379214335236096
     RMSEP = 1.268420270693499